Modifying the MultipleShiftEELSAcquire plugin

icbicket opened the pull request:
- Improve acquisition speed and allow the user more choices
 - Eliminate redundant calls to camera acquisition and unneeded sleep times to reduce the time required to run the plugin.
 - Allow the user to import a dark reference file from disk
 - Allow the user to choose whether or not to apply the dark reference and cross-correlation to the acquired stack
 - Add documentation for the User Guide for MultipleShiftEELSAcquire

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nion-software/nionswift-instrumentation-kit?pullRequest=67) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nion-software/nionswift-instrumentation-kit?pullRequest=67) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nion-software/nionswift-instrumentation-kit?pullRequest=67) it.</sub>

cmeyer commented:
Hi @icbicket thanks for this PR. I will check on the CLA as mentioned in your email.

This looks like it makes some good changes but it will take me a bit of time to review. 

Has this been tested on a live system or just the simulator?

One other comment is that it looks like this was run through a source code formatter. Even though the original format isn't ideal, it makes it difficult to quickly compare changes.

cmeyer reviewed (changes requested):
While waiting for the CLA stuff to finish, it would be helpful to resubmit this PR with the formatting only changes removed. You can submit formatting changes separately in another PR after this one is accepted and merged, although even then we'd prefer not to merge formatting changes unless there is a direct need for them.

cmeyer commented on nionswift_plugin/nion_instrumentation_ui/MultipleShiftEELSAcquire.py:
```diff
@@ -278,21 +448,44 @@ def close(self):
-            AcquireController().start_threaded_acquire_and_sum(stem_controller, eels_camera, number_frames,
-                                                               energy_offset, sleep_time, self.document_controller,
-                                                               functools.partial(self.set_final_layout))
+            if self.dark_ref_choice.checked is False:
```
By the way, force push is ok here, or another commit if you prefer, but there is no need to submit a new PR -- just use this one. I may squash multiple commits if necessary when merging.

icbicket replied:
I've made the changes you suggested, and also fixed another problem I spotted (which I think came from me messing up squashing some commits together earlier): 8aaab4fab560ab5bfb66892f81de8ae3cf634608

icbicket commented:
Thanks for the feedback, @cmeyer . 

To answer your questions, it has been checked on the Rutgers machine, though it was in an earlier form and I've made some small changes to neaten up the code since then, and I've run it successfully in the simulator on my own computer.

I checked the Contribution guidelines on the main repo, and ran it through mypy and flake8 as per that guide, but you have a fair point that it is more difficult to compare versions with these changes.

cmeyer commented:
Ah ok that sounds good then. I'll edit the guidelines about flake8 to making typing/mypy and flake8 option if the existing code isn't already passing those. My thinking was just to ensure that new code passes the checks. So you can decide if you want to back out the formatting or not.

There is the other change to make about moving access to the UI access into the caller and passing as an parameter.

icbicket commented:
Ok, it makes sense to me to have it nicely formatted at the end of the day, but I could push a commit with it unformatted, then another formatted to make it easier for you to check the diffs on the first commit.

I see the other requested change, I need to check/figure out how to do that and I'll come back to it.

cmeyer reviewed (changes requested):
Thanks for the last set of changes. Now I've also had a chance to look at the code more thoroughly. There are two main areas that we request you change.

1) Can you pass `dark_ref_enabled: bool` and `dark_ref_data: typing.Optional[numpy.ndarray]` to `acquire_series` and `start_threaded_acquire_and_sum` (I've included the types, which are useful to have in the function signature)?

Right now `dark_ref` is untyped and can be None, False, or a string, which can be confusing without inspecting the code closely.

And by passing the data instead of the path, we move the file loading up to the UI level which is useful for the next request. This will also require you to move the `read_data_items` call up to the UI level.

2) Next point: we do not want to put a file path into the UI. This is a more difficult problem to solve, so I suggest splitting this PR into two parts. The first part can be this PR exactly but with the line `column.add(dialog_row_df)` commented out. This will allow you to comment it in easily for debugging this PR; but comment it out when submitting the PR. Also, it may require some edits to the userguide too until we figure out the permanent UI.

Once this PR is merged, then we can move on to the UI for the dark image. It is a separate, larger discussion, but this is a problem we'd like to solve for our own acquisition packages too. So we'd like to have a slightly different UI for acquiring the dark image and reusing it that doesn't involve the user entering a path in the UI.

icbicket commented:
I've modified it to do as you suggest with the dark references - keeping the import of the reference inside the ControlView class, and splitting the function inputs into a boolean for the `dark_ref_enabled` choice and a parameter for the `dark_ref_data` (818bb3ead106f5317824faeef3339b3635a52cb5).

I added in typing information for the functions - the ones you highlighted, and some of the other function inputs (901a3fb25dec3b6311529d4bf3f3c664ee3a5056).

For your second suggestion, I understand that you'd like to remove the possibility for the user to input a filepath for now, until the larger discussion about how to store reference images can be resolved, and then later integrate that solution into this plug-in?

cmeyer commented:
> For your second suggestion, I understand that you'd like to remove the possibility for the user to input a filepath for now, until the larger discussion about how to store reference images can be resolved, and then later integrate that solution into this plug-in?

Yes, that's correct. The rest of the changes look great!

icbicket commented:
Got it, I've removed the references to the filepath, and modified the User Guide to match the update to the UI (f8dfeb4). 

If you have any advice on how to do the set-up needed for unit tests, I'd be happy to write some for this plug-in, either in this PR or something to work on for a later one.

cmeyer commented:
This PR has been merged: b7e742c053583c23fc92502485081949ff6fb7de..50933f9de91e8e7636db537e11cc8033b7332fec

cmeyer commented:
@icbicket  FYI to finish your changes (selectable dark references) will require https://github.com/nion-software/nionswift/issues/661 to be implemented next. That's a larger project that we're working on in the next few weeks.

icbicket commented:
Ok, thanks for letting me know, I'll keep an eye on it!
